Mathlib/Tactic/LibrarySearch.lean

```diff
@@ -76,11 +95,19 @@ def librarySearch (goal : MVarId) (lemmas : DiscrTree Name) (solveByElimDepth :=
             newGoal.withContext do
               trace[Tactic.librarySearch] "proving {← addMessageContextFull (mkMVar newGoal)}"
               solveByElim solveByElimDepth newGoal
-          pure $ some $ Sum.inr ()
+          if (← checkRequired) then
+            pure $ some $ Sum.inr ()
+          else
+            set state0
+            pure none
         catch _ =>
           let res := some $ Sum.inl (← getMCtx, newGoals)
-          set state0
-          pure res)
+          if (← checkRequired) then
+            set state0
+            pure res
+          else
+            set state0
+            pure none)
```

Comment:
```suggestion
          set state0
          return if (← checkRequired) then res else none)
```

Reply:
I think this is incorrect, although I'm failing to write a good test showing the behaviour. `checkRequired` should be called before the state is reverted!

e.g. in
```
def using_f (f : α → β) : β := by library_search using f
```
with the current code we only get suggestions that actually use `f` (even though none of them close the goal) while with your change we get suggestions not using `f`.

This is why I left a comment about not reverting the state in the doc-string of `checkRequired`, but perhaps it would be better at this `if` statement.

Reply:
I would change it to `let check ← checkRequired; set state0; return if check then res else none` in that case.